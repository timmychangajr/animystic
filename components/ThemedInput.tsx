import { Keyboard, TextInput, TextInputProps } from 'react-native';

import { AppColors } from '@/constants/Colors';
import { defaultPadding, globalStyles } from '@/constants/globalStyles';

export type ThemedInputProps = TextInputProps & {
  isSafeArea?: boolean;
  backgroundColor?: keyof typeof AppColors;
  withShadow?: boolean;
};

export function ThemedInput({ style, isSafeArea, backgroundColor, withShadow = false, ...otherProps }: ThemedInputProps) {
  return (
    <TextInput
      {...otherProps}
      onBlur={Keyboard.dismiss}
      placeholder={otherProps.placeholder || "dot dot dot"}
      cursorColor={AppColors.text}
      placeholderTextColor={AppColors.quietText}
      style={[
        withShadow && globalStyles.shadowProps,
        {
          // height: 44,
          fontSize: 18,
          fontFamily: 'Outfit',
          paddingHorizontal: defaultPadding,
          borderRadius: 9999,
          width: '100%',
          backgroundColor: AppColors[backgroundColor ?? 'inputBackground'],
          color: AppColors.text,
        }, style
      ]}
    />
  );
}
