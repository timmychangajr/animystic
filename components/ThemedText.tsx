import { AppColors } from '@/constants/Colors';
import { defaultPadding } from '@/constants/globalStyles';
import { StyleSheet, Text, View, type TextProps } from 'react-native';

export type ThemedTextProps = TextProps & {
  textColor?: keyof typeof AppColors;
  type?: 'title' | 'subtitle';
  label?: string;
};

export function ThemedText({
  style,
  textColor,
  type,
  label,
  ...rest
}: ThemedTextProps) {
  const color = AppColors[textColor ?? 'text'];

  return rest.children && (
    <View style={{ flexDirection: 'row', alignItems: 'center', flexWrap: 'wrap' }}>
      {label ? <Text style={{
        color: AppColors.quietText,
        marginRight: defaultPadding,
        fontFamily: 'Outfit',
      }}>{label}</Text> : null}
      <Text
        {...rest}
        style={[
          { color, fontFamily: 'Outfit' },
          styles.default,
          type === 'title' ? styles.title : undefined,
          type === 'subtitle' ? styles.subtitle : undefined,
          style,
        ]}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  default: {
    fontSize: 16,
    lineHeight: 24,
  },
  title: {
    fontSize: 25,
    lineHeight: 32,
  },
  subtitle: {
    fontSize: 20,
  },
});
