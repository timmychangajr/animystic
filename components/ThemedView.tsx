import { View, type ViewProps } from 'react-native';

import { AppColors } from '@/constants/Colors';
import { globalStyles } from '@/constants/globalStyles';
import { SafeAreaView } from 'react-native-safe-area-context';

export type ThemedViewProps = ViewProps & {
  isSafeArea?: boolean;
  backgroundColor?: keyof typeof AppColors;
  withShadow?: boolean;
};

export function ThemedView({ style, isSafeArea, backgroundColor, withShadow = false, ...otherProps }: ThemedViewProps) {
  const MainView = isSafeArea ? SafeAreaView : View;
  return (
    <MainView
      {...otherProps}
      style={[
        withShadow && globalStyles.shadowProps,
        {
          backgroundColor: AppColors[backgroundColor ?? 'background']
        }, style
      ]}
    />
  );
}
