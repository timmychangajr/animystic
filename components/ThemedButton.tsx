import { ActivityIndicator, TouchableOpacity, TouchableOpacityProps } from 'react-native';

import { AppColors } from '@/constants/Colors';
import { defaultPadding, globalStyles } from '@/constants/globalStyles';
import { Octicons } from '@expo/vector-icons';
import { ThemedText, ThemedTextProps } from './ThemedText';

export type ThemedButtonProps = TouchableOpacityProps & {
  isSafeArea?: boolean;
  isLoading?: boolean;
  title?: string;
  textType?: ThemedTextProps['type'];
  backgroundColor?: keyof typeof AppColors;
  withShadow?: boolean;
};

export function ThemedButton({
  style,
  isSafeArea,
  backgroundColor,
  withShadow = false,
  isLoading = false,
  textType = 'subtitle',
  title = 'Submit',
  ...otherProps }: ThemedButtonProps) {
  return (
    <TouchableOpacity
      {...otherProps}
      disabled={isLoading}
      style={[
        withShadow && globalStyles.shadowProps,
        {
          backgroundColor: AppColors[backgroundColor ?? 'buttonBackground'],
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'center',
          margin: 'auto',
          borderRadius: 9999,
          gap: defaultPadding / 2,
          paddingHorizontal: defaultPadding,
          paddingVertical: defaultPadding / 2.5,
        }, style
      ]}
    >
      {isLoading ? <ActivityIndicator size='small' color={AppColors.text} /> :
        <>
          <Octicons name='search' color={AppColors.text} size={20} />
          <ThemedText type={textType}>{title}</ThemedText>
        </>
      }
    </TouchableOpacity>
  );
}
