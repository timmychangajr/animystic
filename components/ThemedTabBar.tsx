import { AppColors } from '@/constants/Colors';
import { defaultPadding } from '@/constants/globalStyles';
import * as Haptics from 'expo-haptics';
import { Octicons } from '@expo/vector-icons';
import { BottomTabBarProps } from '@react-navigation/bottom-tabs';
import { StyleSheet, TouchableOpacity, View } from 'react-native';
import { ThemedView } from './ThemedView';

interface Route {
  icon: keyof typeof Octicons.glyphMap,
  name: string,
}

const routeDetail: Record<string, Route> = {
  index: { name: 'Home', icon: 'home' },
  details: { name: 'Details', icon: 'log' }
}
export function ThemedTabBar({ state, navigation }: BottomTabBarProps) {
  return (
    <ThemedView style={styles.container}>
      <View style={{ flexDirection: 'row', gap: defaultPadding*3 }}>
        {state.routes.map((route, index) => {
          const isFocused = state.index === index;
          const showTab = route.name === 'index' || (route.name === 'details' && !!route.params)

          const onPress = () => {
            Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Soft);
            const event = navigation.emit({
              type: 'tabPress',
              target: route.key,
              canPreventDefault: true,
            });

            if (!isFocused && !event.defaultPrevented) {
              navigation.navigate(route.name, route.params);
            }
          };

          return showTab && (
            <TouchableOpacity key={route.key} style={styles.tabButton} onPress={onPress}>
              <Octicons name={routeDetail[route.name].icon} size={isFocused? 22: 20} color={AppColors.icon} />
              <View
                style={[
                  styles.indicator,
                  isFocused && styles.tabButtonSelected
                ]} />
            </TouchableOpacity>
          )
        })}
      </View>
    </ThemedView>
  )
}

const styles = StyleSheet.create({
  container: {
    paddingBottom: 44,
    gap: 10,
    justifyContent: 'center',
    alignItems: 'center',
  },
  tabButton: {
    borderBottomColor: 'transparent',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  indicator: {
    position: 'absolute',
    bottom: -defaultPadding/2,
    height: 4,
    width: '70%',
    borderRadius: 999,
  },
  tabButtonSelected: {
    backgroundColor: '#fff'
  },
  topBorder: {
    height: 3,
    margin: 'auto',
    width: '70%',
    borderRadius: 10,
    opacity: 0.2,
  },
  separator: {
    height: '100%',
    backgroundColor: 'transparent',
    width: 3,
    borderRadius: 10,
    marginHorizontal: 40,
    opacity: 0.2,
  }
})
