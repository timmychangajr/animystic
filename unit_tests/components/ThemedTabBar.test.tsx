import { ThemedTabBar, ThemedTabBarProps } from '@/components/ThemedTabBar';
import { cleanup, fireEvent, render } from '@testing-library/react-native';

describe('ThemedTabBar', () => {
  afterEach(cleanup);
  let mockProps: ThemedTabBarProps = {
    state: {
      key: 'string',
      index: 0,
      routeNames: [],
      routes: [
        { name: 'index', key: 'index_123', params: {} },
        { name: 'details', key: 'details_123', params: undefined }
      ],
    },
    navigation: {
      emit: jest.fn(jest.fn(() => ({ defaultPrevented: false }))),
      navigate: jest.fn()
    },
  }
  test('renders correctly', async () => {
    const tree = render(<ThemedTabBar {...mockProps} />);
    expect(tree).toMatchSnapshot();
  });
  test('should load index as only tab', async () => {
    const tree = render(<ThemedTabBar {...mockProps} />);
    expect(tree.queryByTestId('tab_index')).toBeTruthy()
    expect(tree.queryByTestId('tab_details')).toBeNull()
  })
  test('should allow navigation to details when params are present', async () => {
    const newMockProps = {
      ...mockProps,
      state: {
        ...mockProps.state,
        routes: [
          { name: 'index', key: 'index_123', params: {} },
          { name: 'details', key: 'details_123', params: {} }
        ],
      }
    }
    const tree = render(<ThemedTabBar {...newMockProps} />);
    const tabButton = tree.getByTestId('tab_details')
    fireEvent.press(tabButton)
  })
});