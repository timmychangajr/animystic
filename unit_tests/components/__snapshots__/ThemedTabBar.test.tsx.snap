// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemedTabBar renders correctly 1`] = `
<View
  style={
    [
      false,
      {
        "backgroundColor": "#000000B5",
      },
      {
        "alignItems": "center",
        "gap": 10,
        "justifyContent": "center",
        "paddingBottom": 44,
      },
    ]
  }
>
  <View
    style={
      {
        "flexDirection": "row",
        "gap": 60,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderBottomColor": "transparent",
          "flexDirection": "row",
          "justifyContent": "center",
          "opacity": 1,
        }
      }
      testID="tab_index"
    >
      <Text />
      <View
        style={
          [
            {
              "borderRadius": 999,
              "bottom": -10,
              "height": 4,
              "position": "absolute",
              "width": "70%",
            },
            {
              "backgroundColor": "#fff",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
